### PR TITLE
*: refactor split check

### PR DESCRIPTION
**Pull request by overvenus**
Move size split check from SplitCheckRunner to CoprocessorHost. It also move `region-max-size` and `region-split-size` to coprocessor configuration.

Ref #2378

### CODE REVIEW

File: tests/config/test-custom-latest.toml [hunk 1]
```diff
@@ -73,6 +71,10 @@ raft-store-max-leader-lease = "12s"
 right-derive-when-split = false
 allow-remove-leader = true
 
+[coprocessor]
+region-max-size = "12MB"
```
Comment (siddontang) [inline tests/config/test-custom-latest.toml]: @LinuxGit 

Do we need to update ansible?

**Comment by siddontang** [comment]
No need to support compatible config file.

File: etc/config-template.toml [hunk 2]
```diff
@@ -138,6 +133,16 @@
 # Interval (s) to check region whether the data are consistent.
 # consistency-check-interval = 0
 
+[coprocessor]
+# When it is true, it will try to split a region with table prefix if
+# that region crosses tables.
+# split-region-on-table = false
```
Comment (BusyJay) [inline etc/config-template.toml]: Stale comment.

File: etc/config-template.toml [hunk 3]
```diff
@@ -95,11 +95,6 @@
 # Store heartbeat tick interval for reporting to pd.
 # pd-store-heartbeat-tick-interval = "10s"
 
-# When the region's size exceeds region-max-size, we will split the region
-# into two which the left region's size will be region-split-size or a little
-# bit smaller.
-# region-max-size = "144MB"
```
Comment (BusyJay) [inline etc/config-template.toml]: If this is set, then `coprocessor.region-max-size` should be updated too.

File: src/bin/tikv-server.rs [hunk 4]
```diff
@@ -463,6 +463,7 @@ fn main() {
     if let Err(e) = config.validate() {
         fatal!("invalid configuration: {:?}", e);
     }
+    config.compatible_adjust();
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: Should adjust before validating.

File: src/raftstore/store/config.rs [hunk 5]
```diff
@@ -202,6 +200,15 @@ impl Config {
             ));
         }
 
+        // Compatibility check.
+        if self.region_max_size.0 != 0 {
+            warn!("raftstore.region-max-size has been moved to coprocessor");
```
Comment (BusyJay) [inline src/raftstore/store/config.rs]: This is unreachable.

File: src/raftstore/store/keys.rs [hunk 6]
```diff
@@ -195,7 +195,7 @@ pub fn data_key(key: &[u8]) -> Vec<u8> {
 }
 
 pub fn origin_key(key: &[u8]) -> &[u8] {
-    assert!(validate_data_key(key));
+    assert!(validate_data_key(key), "invalid data key {:?}", key);
```
Comment (BusyJay) [inline src/raftstore/store/keys.rs]: How about `util::escape`?

**Comment by siddontang** [comment]
LGTM

**Comment by BusyJay** [comment]
LGTM

File: etc/config-template.toml [hunk 7]
```diff
@@ -138,6 +133,13 @@
 # Interval (s) to check region whether the data are consistent.
 # consistency-check-interval = 0
 
+[coprocessor]
```
Comment (zhangjinpeng1987) [inline etc/config-template.toml]: I think this will confuse ops.

**Pull request merged**